gapic/samplegen/samplegen.py

```diff
@@ -102,6 +104,94 @@ class TransformedRequest:
     base: str
     single: Optional[AttributeRequestSetup]
     body: Optional[List[AttributeRequestSetup]]
+    pattern: Optional[str] = None
+
+    # Resource patterns look something like
+    # kingdom/{kingdom}/phylum/{phylum}/class/{class}
+    RESOURCE_RE = re.compile(r"\{([^}/]+)\}")
+
+    @classmethod
+    def build(cls, request_type: wrappers.MessageType, api_schema, base: str,
+              attrs: List[AttributeRequestSetup], is_resource_request: bool):
+        """Build a TransformedRequest based on parsed input.
+
+        Acts as a factory to hide complicated logic for resource based requests.
+
+        Args:
+            request_type (wrappers.MessageType): The method's request message type.
+            api_schema (api.API): The API schema (used for looking up other messages)
+            base (str): the name of the base field being set.
+            attrs List[str]: All the attributes (or fields) being set
+                                           for the base field.
+            is_resource_request (bool): Indicates whether the request describes a
+                                        constructed resource name path.
+
+        Returns:
+            TransformedRequest
+
+        Raises:
+            NoSuchResource: If the base parameter field for a resource-name
+                            request statement lists a resource_type for which
+                            there is no message with the same resource type.
+            NoSuchResourcePattern: If all the request setup statements for a
+                                   resource name parameter do not combine to
+                                   match a valid path pattern for that resource.
+        """
+
+        if not attrs[0].field:
```

Comment:
not an expert on proto internals, so this might be a silly question: `attrs[0]` always exists?

Reply:
Not a silly question, but also not a proto-related question: in `validate_and_transform_request`, a map between an attribute base and a list of AttributeRequestSetup objects gets created; during the construction of the map, every non-error lookup results in an append to the corresponding list. When the TransformedRequests are built, they are built from the key/value pairs in that map, so every attr list is guaranteed to have at least one element.
I'll add a comment to this effect.